From 6fe40928ee94475cacf3479a6d8e78a74d70163a Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 19 Jan 2024 10:53:52 +0100
Subject: [PATCH 0639/1695] plat-stm32mp2: add hse monitoring feature

-setup hse monitoring while boot_final level
-register stm32_hse_over_frequency handler on lptimer alarm

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d69b265f4fab41f5312b530f6a7815c690c4f9ee plat-stm32mp2: add hse monitoring feature]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I33eb8d8b48082ab954918ce2426552ec5d9d6219
---
 core/arch/arm/plat-stm32mp2/main.c | 74 ++++++++++++++++++++++++++++++
 1 file changed, 74 insertions(+)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 6ce096ed3..7df2efe74 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -5,6 +5,8 @@
 
 #include <config.h>
 #include <console.h>
+#include <drivers/clk_dt.h>
+#include <drivers/counter.h>
 #include <drivers/gic.h>
 #include <drivers/rstctrl.h>
 #include <drivers/stm32_bsec.h>
@@ -256,3 +258,75 @@ void __noreturn do_reset(const char *str __maybe_unused)
 	/* Can't occur */
 	panic();
 }
+
+#ifdef CFG_STM32_HSE_MONITORING
+/* pourcent rate of hse alarm */
+#define HSE_ALARM_PERCENT	110
+#define FREQ_MONITOR_COMPAT	"st,freq-monitor"
+
+static void stm32_hse_over_frequency(uint32_t ticks __unused,
+				     void *user_data __unused)
+{
+	EMSG("HSE over frequency: nb ticks:%"PRIu32, ticks);
+}
+DECLARE_KEEP_PAGER(stm32_hse_over_frequency);
+
+static TEE_Result stm32_hse_monitoring(void)
+{
+	struct counter_device *counter = NULL;
+	struct clk *hse_clk = NULL;
+	struct clk *hsi_clk = NULL;
+	unsigned long hse = 0;
+	unsigned long hsi_cal = 0;
+	uint32_t ticks = 0;
+	void *fdt = NULL;
+	void *config = NULL;
+	int node = 0;
+	int res  = 0;
+
+	DMSG("HSE monitoring");
+
+	fdt = get_embedded_dt();
+	node = fdt_node_offset_by_compatible(fdt, 0, FREQ_MONITOR_COMPAT);
+	if (node < 0)
+		panic();
+
+	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return TEE_SUCCESS;
+
+	res = clk_dt_get_by_name(fdt, node, "hse", &hse_clk);
+	if  (res)
+		return res;
+
+	res = clk_dt_get_by_name(fdt, node, "hsi", &hsi_clk);
+	if  (res)
+		return res;
+
+	hse = clk_get_rate(hse_clk);
+	hsi_cal = clk_get_rate(hsi_clk);
+
+	/*
+	 * hsi_cal is based on hsi & DIVISOR
+	 * DIVISOR is fixed (1024)
+	 */
+	hsi_cal /= 1024;
+
+	ticks = (hse / 100) * HSE_ALARM_PERCENT;
+	ticks /= hsi_cal;
+
+	DMSG("HSE:%luHz HSI cal:%luHz alarm:%"PRIu32, hse, hsi_cal, ticks);
+
+	counter = fdt_counter_get(fdt, node, &config);
+	assert(counter && config);
+
+	counter->alarm.callback = stm32_hse_over_frequency;
+	counter->alarm.ticks = ticks;
+
+	counter_start(counter, config);
+	counter_set_alarm(counter);
+
+	return TEE_SUCCESS;
+}
+
+driver_init_late(stm32_hse_monitoring);
+#endif /* CFG_STM32_HSE_MONITORING */
-- 
2.47.2

